components/script/dom/node.rs

```diff
@@ -1283,37 +1258,22 @@ impl Iterator for PrecedingNodeIterator {
     }
 }
 
-pub struct LastChildIterator {
-    current: Option<Root<Node>>,
-}
-
-impl Iterator for LastChildIterator {
-    type Item = Root<Node>;
-
-    fn next(&mut self) -> Option<Root<Node>> {
-        let current = match self.current.take() {
-            None => return None,
-            Some(current) => current,
-        };
-        self.current = current.GetLastChild();
-        Some(current)
-    }
-}
-
-pub struct AncestorIterator {
+struct SimpleNodeIterator<I>
+    where I: Fn(&Node) -> Option<Root<Node>>
+{
     current: Option<Root<Node>>,
+    next_node: I,
 }
 
-impl Iterator for AncestorIterator {
+impl<I> Iterator for SimpleNodeIterator<I>
+    where I: Fn(&Node) -> Option<Root<Node>>
+{
     type Item = Root<Node>;
 
-    fn next(&mut self) -> Option<Root<Node>> {
-        let current = match self.current.take() {
-            None => return None,
-            Some(current) => current,
-        };
-        self.current = current.GetParentNode();
-        Some(current)
+    fn next(&mut self) -> Option<Self::Item> {
+        let current = self.current.take();
+        self.current = current.as_ref().and_then(|c| (&self.next_node)(c));
```

Comment:
the `&` was unnecessary. i removed it in the latest force push.

regarding the parens. if i wrote:

```
self.next_node(c)
```

here's the error i get from the compiler:

```
error[E0599]: no method named `next_node` found for type `&mut dom::node::SimpleNodeIterator<I>` in the current scope
    --> /Users/corey/dev/servo/components/script/dom/node.rs:1275:59
     |
1275 |         self.current = current.as_ref().and_then(|c| self.next_node(c));
     |                                                           ^^^^^^^^^ field, not a method
     |
     = help: use `(self.next_node)(...)` if you meant to call the function stored in the `next_node` field
```

so presumably, rust is looking for a _method_ on the `SimpleNodeIterator` structure called `next_node`, but since there isn't a method called `next_node`, we get a compiler error. we need to explicitly indicate we want to retrieve the function stored as a field, and then call that.